ads/google/a4a/performance.js

```diff
@@ -194,36 +198,65 @@ export class GoogleAdLifecycleReporter extends BaseLifecycleReporter {
    * @override
    */
   sendPing(name) {
-    this.emitPing_(this.buildPingAddress_(name));
+    this.emitPing_(this.buildPingAddress_(name, this.extraVariables_));
   }
 
   /**
+   * @param {string} variable
+   * @param {string|number} value
+   * @override
+   */
+  setPingVariable(variable, value) {
+    this.extraVariables_[variable] = value;
+  }
+
+  /**
+   * @return {number}
+   */
+  getSlotId() {
+    return this.slotId_;
+  }
+
+  /** @override */
+  reset() {
+    this.extraVariables_ = new Object(null);
+  }
+
+
+  /**
    * @param {string} name  Metric name to send.
+   * @param {Object<string, string|number>=} opt_extraParams
    * @returns {string}  URL to send metrics to.
    * @private
    */
-  buildPingAddress_(name) {
+  buildPingAddress_(name, opt_extraParams) {
     const stageId = LIFECYCLE_STAGES[name] || 9999;
     const delta = Date.now() - this.initTime_;
-    // Note: QQid comes from a network header and eid could, potentially, be
-    // injected by a publisher.  Treat both of them as unverified user content
-    // and encode before inserting them into URI.
-    const encodedQqid = this.qqid_ ?
-        encodeURIComponent(this.qqid_) : false;
-    const qqidParam = encodedQqid ?
-        `&qqid.${this.slotId_}=${encodedQqid}` : '';
-    const eid = this.element_.getAttribute(EXPERIMENT_ATTRIBUTE);
-    const eidParam = eid ? `&e=${encodeURIComponent(eid)}` : '';
+    let extraParams = '';
+    if (opt_extraParams) {
+      const paramList = [];
+      for (const param in opt_extraParams) {
+        if (!!param) {
+          paramList.push(
+              encodeURIComponent(param) + '=' +
+              encodeURIComponent(opt_extraParams[param]));
+        }
+      }
+      if (paramList.length > 0) {
+        extraParams = '&' +
+            this.urlReplacer_.expandStringSync(paramList.join('&'));
```

Comment:
Why sync?  If async, we could get a wider set of allowed expansion macros.  There's also the possibility that a key given in the extra params could itself get expanded here... not sure how relevant that is but you could guard against it by only expanding values.